File: client-sdk/ios/WebRTCKit/WebRTC/ReconnectionManager.swift
// MARK: - Reconnection Manager
// TRUST 5 Compliance: Testable, Unified, Trackable

import Foundation
import WebRTC
import Combine

// MARK: - Reconnection Manager

/// Manager for automatic reconnection with state machine
/// Handles reconnection states: Stable, Reconnecting, Failed
/// Implements exponential backoff and ICE restart logic
@available(iOS 13.0, *)
final class ReconnectionManager {

    // MARK: - Properties

    @Published private(set) var reconnectionState: ReconnectionState = .stable
    @Published private(set) var retryCount: Int = 0

    private let mutex = NSLock()
    private var currentAttempt: ReconnectionAttempt?

    private enum Constants {
        static let maxRetryAttempts = 3
        static let backoffDelays: [TimeInterval] = [1.0, 2.0, 4.0] // 1s, 2s, 4s
    }

    // MARK: - Nested Types

    /// Reconnection states
    enum ReconnectionState {
        case stable       // Connection is stable
        case reconnecting // Attempting to reconnect
        case failed       // All reconnection attempts failed
    }

    /// Reconnection failure types
    enum FailureType {
        case minor   // ICE restart can fix (e.g., candidate pair failure)
        case major   // Full reconnection needed (e.g., peer connection closed)
        case fatal   // Cannot recover (e.g., authentication failure)
    }

    /// Reconnection strategies
    enum ReconnectionStrategy {
        case iceRestart        // Restart ICE only (keep peer connection)
        case fullReconnection  // Full reconnection (create new peer connection)
    }

    private struct ReconnectionAttempt {
        let attemptNumber: Int
        let failureType: FailureType
        let timestamp: Date
    }

    // MARK: - Public Methods

    /// Handle connection failure and attempt reconnection
    /// - Parameters:
    ///   - failureType: Type of failure that occurred
    ///   - onReconnect: Callback to perform reconnection
    /// - Returns: Result of reconnection attempt
    func handleFailure(
        failureType: FailureType,
        onReconnect: @escaping (ReconnectionStrategy) async throws -> Void
    ) async {
        mutex.lock()

        // Check if already reconnecting
        guard reconnectionState != .reconnecting else {
            mutex.unlock()
            return
        }

        // Check if max attempts reached
        guard retryCount < Constants.maxRetryAttempts else {
            reconnectionState = .failed
            mutex.unlock()
            return
        }

        // Update state to reconnecting
        reconnectionState = .reconnecting
        let attemptNumber = retryCount + 1
        currentAttempt = ReconnectionAttempt(
            attemptNumber: attemptNumber,
            failureType: failureType,
            timestamp: Date()
        )
        retryCount = attemptNumber
        mutex.unlock()

        // Determine reconnection strategy
        let strategy: ReconnectionStrategy
        switch failureType {
        case .minor:
            strategy = .iceRestart
        case .major:
            strategy = .fullReconnection
        case .fatal:
            reconnectionState = .failed
            return
        }

        // Get backoff delay
        let backoffDelay = getCurrentBackoffDelay()

        // Wait before attempting reconnection
        try? await Task.sleep(nanoseconds: UInt64(backoffDelay * 1_000_000_000))

        // Attempt reconnection
        do {
            try await onReconnect(strategy)

            // Reconnection successful
            mutex.lock()
            reconnectionState = .stable
            retryCount = 0
            currentAttempt = nil
            mutex.unlock()

        } catch {
            // Reconnection failed
            mutex.lock()

            if attemptNumber >= Constants.maxRetryAttempts {
                reconnectionState = .failed
            } else {
                // Stay in reconnecting state for next attempt
                reconnectionState = .reconnecting
            }

            mutex.unlock()
        }
    }

    /// Reset reconnection state (call when connection is stable)
    func reset() {
        mutex.lock()
        reconnectionState = .stable
        retryCount = 0
        currentAttempt = nil
        mutex.unlock()
    }

    /// Get current backoff delay for logging
    func getCurrentBackoffDelay() -> TimeInterval {
        let attempt = retryCount
        if attempt <= Constants.backoffDelays.count {
            return Constants.backoffDelays[attempt - 1]
        } else {
            let multiplier = 1 << (attempt - Constants.backoffDelays.count)
            return Constants.backoffDelays.last! * TimeInterval(multiplier)
        }
    }

    /// Check if reconnection is possible
    func canReconnect() -> Bool {
        mutex.lock()
        let canRetry = retryCount < Constants.maxRetryAttempts &&
                       reconnectionState != .failed
        mutex.unlock()
        return canRetry
    }
}

File: client-sdk/android/webrtc-core/src/main/java/com/webrtclite/core/webrtc/ReconnectionManager.kt
package com.webrtclite.core.webrtc

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.delay
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.webrtc.PeerConnection
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Manager for automatic reconnection with state machine
 * Handles reconnection states: Stable, Reconnecting, Failed
 * Implements exponential backoff and ICE restart logic
 */
@Singleton
class ReconnectionManager @Inject constructor() {

    // State management
    private val _reconnectionState = MutableStateFlow(ReconnectionState.STABLE)
    val reconnectionState: StateFlow<ReconnectionState> = _reconnectionState.asStateFlow()

    private val _retryCount = MutableStateFlow(0)
    val retryCount: StateFlow<Int> = _retryCount.asStateFlow()

    private val mutex = Mutex()

    companion object {
        private const val MAX_RETRY_ATTEMPTS = 3
        private val BACKOFF_DELAYS = listOf(1000L, 2000L, 4000L) // 1s, 2s, 4s
    }

    /**
     * Reconnection states
     */
    enum class ReconnectionState {
        STABLE,      // Connection is stable
        RECONNECTING, // Attempting to reconnect
        FAILED        // All reconnection attempts failed
    }

    /**
     * Reconnection failure types
     */
    enum class FailureType {
        MINOR,   // ICE restart can fix (e.g., candidate pair failure)
        MAJOR,   // Full reconnection needed (e.g., peer connection closed)
        FATAL    // Cannot recover (e.g., authentication failure)
    }

    private data class ReconnectionAttempt(
        val attemptNumber: Int,
        val failureType: FailureType,
        val timestamp: Long = System.currentTimeMillis()
    )

    private var currentAttempt: ReconnectionAttempt? = null

    /**
     * Handle connection failure and attempt reconnection
     * @param failureType Type of failure that occurred
     * @param peerConnection Current peer connection (for ICE restart)
     * @param onReconnect Callback to perform reconnection
     */
    suspend fun handleFailure(
        failureType: FailureType,
        peerConnection: PeerConnection? = null,
        onReconnect: suspend (ReconnectionStrategy) -> Result<Unit>
    ) {
        mutex.withLock {
            if (_reconnectionState.value == ReconnectionState.RECONNECTING) {
                // Already reconnecting, skip
                return
            }

            if (_retryCount.value >= MAX_RETRY_ATTEMPTS) {
                _reconnectionState.value = ReconnectionState.FAILED
                return
            }

            _reconnectionState.value = ReconnectionState.RECONNECTING
            val attemptNumber = _retryCount.value + 1
            currentAttempt = ReconnectionAttempt(attemptNumber, failureType)
            _retryCount.value = attemptNumber
        }

        // Get backoff delay
        val backoffDelay = if (attemptNumber <= BACKOFF_DELAYS.size) {
            BACKOFF_DELAYS[attemptNumber - 1]
        } else {
            BACKOFF_DELAYS.last() * (1L shl (attemptNumber - BACKOFF_DELAYS.size))
        }

        // Wait before attempting reconnection
        delay(backoffDelay)

        // Determine reconnection strategy
        val strategy = when (failureType) {
            FailureType.MINOR -> ReconnectionStrategy.ICE_RESTART
            FailureType.MAJOR -> ReconnectionStrategy.FULL_RECONNECTION
            FailureType.FATAL -> {
                _reconnectionState.value = ReconnectionState.FAILED
                return
            }
        }

        // Attempt reconnection
        val result = onReconnect(strategy)

        if (result.isSuccess) {
            // Reconnection successful
            _reconnectionState.value = ReconnectionState.STABLE
            _retryCount.value = 0
            currentAttempt = null
        } else {
            // Reconnection failed, check if we should retry
            if (attemptNumber >= MAX_RETRY_ATTEMPTS) {
                _reconnectionState.value = ReconnectionState.FAILED
            } else {
                // Stay in reconnecting state for next attempt
                _reconnectionState.value = ReconnectionState.RECONNECTING
            }
        }
    }

    /**
     * Reset reconnection state (call when connection is stable)
     */
    suspend fun reset() {
        mutex.withLock {
            _reconnectionState.value = ReconnectionState.STABLE
            _retryCount.value = 0
            currentAttempt = null
        }
    }

    /**
     * Get current backoff delay for logging
     */
    fun getCurrentBackoffDelay(): Long {
        val attempt = _retryCount.value
        return if (attempt <= BACKOFF_DELAYS.size) {
            BACKOFF_DELAYS.getOrElse(attempt - 1) { BACKOFF_DELAYS.last() }
        } else {
            BACKOFF_DELAYS.last() * (1L shl (attempt - BACKOFF_DELAYS.size))
        }
    }

    /**
     * Check if reconnection is possible
     */
    fun canReconnect(): Boolean {
        return _retryCount.value < MAX_RETRY_ATTEMPTS &&
               _reconnectionState.value != ReconnectionState.FAILED
    }
}

/**
 * Reconnection strategies
 */
enum class ReconnectionStrategy {
    ICE_RESTART,        // Restart ICE only (keep peer connection)
    FULL_RECONNECTION   // Full reconnection (create new peer connection)
}
